apps/hash-frontend/src/shared/keyboard-shortcuts-context.tsx

```diff
@@ -0,0 +1,173 @@
+import {
+  createContext,
+  FunctionComponent,
+  PropsWithChildren,
+  useCallback,
+  useContext,
+  useEffect,
+  useMemo,
+  useState,
+} from "react";
+
+export type KeyboardShortcut = {
+  keys: string[];
+  callback: (event: KeyboardEvent) => void;
+};
+
+type SetKeyboardShortcutsFunction = (shortcuts: KeyboardShortcut[]) => void;
+
+type UnsetKeyboardShortcutsFunction = (
+  shortcutsToUnset: { keys: string[] }[],
+) => void;
+
+type KeyboardShortcutsContextValue = {
+  setKeyboardShortcuts: SetKeyboardShortcutsFunction;
+  unsetKeyboardShortcuts: UnsetKeyboardShortcutsFunction;
+};
+
+export const KeyboardShortcutsContext =
+  createContext<null | KeyboardShortcutsContextValue>(null);
+
+const modKeyToKeyboardEventProperty: Record<string, keyof KeyboardEvent> = {
+  Alt: "altKey",
+  Meta: "metaKey",
+  Control: "ctrlKey",
+  Shift: "shiftKey",
+};
+
+/**
+ * We detect which keys are pressed based on a single KeyboardEvent.
+ *
+ * See limitations described on {@link useSetKeyboardShortcuts}
+ */
+const areAllKeysPressed = (event: KeyboardEvent, keys: string[]) => {
+  for (const key of keys) {
+    const modifierProperty = modKeyToKeyboardEventProperty[key];
+
+    if (event.key !== key && (!modifierProperty || !event[modifierProperty])) {
+      return false;
+    }
+  }
+
+  return true;
+};
+
+export const KeyboardShortcutsContextProvider: FunctionComponent<
+  PropsWithChildren
+> = ({ children }) => {
+  const [keyboardShortcutsState, setKeyboardShortcutsState] = useState<
+    KeyboardShortcut[]
+  >([]);
+
+  const setKeyboardShortcuts: SetKeyboardShortcutsFunction = useCallback(
+    (shortcutsToRegister) => {
+      setKeyboardShortcutsState((currentShortcuts) => {
+        const currentShortcutsWithoutDuplicates = currentShortcuts.filter(
+          (existingShortcut) =>
+            !shortcutsToRegister.some(
+              (newShortcut) =>
+                newShortcut.keys.length === existingShortcut.keys.length &&
+                existingShortcut.keys.every((key) =>
+                  newShortcut.keys.includes(key),
+                ),
+            ),
+        );
+
+        return [...currentShortcutsWithoutDuplicates, ...shortcutsToRegister];
+      });
+    },
+    [],
+  );
+
+  const unsetKeyboardShortcuts: UnsetKeyboardShortcutsFunction = useCallback(
+    (shortcutsKeysToRemove) => {
+      setKeyboardShortcutsState((currentShortcuts) =>
+        currentShortcuts.filter(
+          (currentShortcut) =>
+            !shortcutsKeysToRemove.some(
+              ({ keys: keysToRemove }) =>
+                currentShortcut.keys.length !== keysToRemove.length ||
+                !currentShortcut.keys.every((key) =>
+                  keysToRemove.includes(key),
+                ),
+            ),
+        ),
+      );
+    },
+    [],
+  );
```

Comment:
If I'm understanding correctly, an additional limitation would be that if we have components which override global shortcuts from the command bar for example, then when the component is unmounted and `unsetKeyboardShortcuts` is called the global shortcut won't be restored. But don't think this is worth addressing until we have a use-case for this, but maybe worth keeping in mind or documenting somewhere.

Reply:
I added a comment in https://github.com/hashintel/hash/pull/3603/commits/e3565b2319f575f2373d940d5f2fb38b6217f121